### PR TITLE
spec(core): introduces a page specific for consensus params

### PR DESCRIPTION
**Pull request by cason**
Solves #2413.

Added [`spec/core/params.md`](https://github.com/cometbft/cometbft/blob/cason/2413-spec-params/spec/core/params.md).

Work in progress.

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [ ] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: docs/references/architecture/adr-112-proposer-based-timestamps.md
Comment (adizere) [inline docs/references/architecture/adr-112-proposer-based-timestamps.md:1]: 👍

File: proto/cometbft/types/v1/params.proto [hunk 1]
```diff
@@ -21,55 +21,62 @@ message ConsensusParams {
...
+  // The recommended value of is should correspond to the application's
+  // "unbonding period" or other similar mechanism for handling
+  // Nothing-At-Stake attacks.
+  // See: https://github.com/ethereum/wiki/wiki/Proof-of-Stake-FAQ#what-is-the-nothing-at-stake-problem-and-how-can-it-be-fixed.
   google.protobuf.Duration max_age_duration = 2
       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
 
-  // This sets the maximum size of total evidence in bytes that can be committed in a single block.
-  // and should fall comfortably under the max block bytes.
-  // Default is 1048576 or 1MB
```
Comment (adizere) [inline proto/cometbft/types/v1/params.proto:54]: We should not mention the default for `max_bytes` anymore?

File: proto/cometbft/types/v1/params.proto [hunk 2]
```diff
@@ -81,39 +88,45 @@ message HashedParams {
...
 
-// SynchronyParams configure the bounds under which a proposed block's timestamp is considered valid.
-// These parameters are part of the proposer-based timestamps algorithm. For more information,
-// see the specification of proposer-based timestamps:
-// https://github.com/cometbft/cometbft/tree/main/spec/consensus/proposer-based-timestamp
+// SynchronyParams determine the validity of block timestamps.
+// 
+// These parameters are part of the Proposer-Based Timestamps (PBTS) algorithm.
+// For more information on the relationship of the synchrony parameters to
+// block timestamps validity, refer to the PBTS specification:
```
Comment (adizere) [inline proto/cometbft/types/v1/params.proto:95]:
```suggestion
// block timestamps validity, refer to the PBTS specification: https://github.com/cometbft/cometbft/tree/main/spec/consensus/proposer-based-timestamp
```

File: spec/consensus/time.md [hunk 3]
```diff
@@ -16,15 +16,19 @@ The Time produced by CometBFT satisfies the following properties:
...
+  to arbitrarily increase or decrease the block Time.
+  In other words, the Time of blocks should be defined by correct nodes.
 
 In addition, the Time produced by CometBFT is expected, by external observers, to provide:
 
 - **Relation to real time**: block times bear some resemblance to real time.
   In other words, block times should represent, within some reasonable accuracy,
-  the actual time at which blocks were produced.
+  the actual clock time at which blocks were produced.
+  More formally, lets `t` be the clock time at which a block with header `H`
```
Comment (adizere) [inline spec/consensus/time.md:28]:
```suggestion
  More formally, let `t` be the clock time at which a block with header `H`
```

nit

File: spec/consensus/time.md [hunk 4]
```diff
@@ -33,20 +37,53 @@ CometBFT implements two algorithms for computing block times:
...
-------------|-------------------|-----------------------------|-----------------------|
-| BFT Time  | Guaranteed        | Tolerates `< 1/3` Byzantine nodes     | Best effort and **not** guaranteed |
-| PBTS      | Guaranteed        | Tolerates `< 2/3` Byzantine nodes     | Guaranteed within configured synchronous parameters: `PRECISION` and `MSGDELAY` |
+The table below compares BFT Time and PBTS algorithms in terms of the above enumerated properties:
+
+| Algorithm | Time Monotonicity | Byzantine Fault Tolerance         | Relation to real time                                                                         |
+|-----------|:-----------------:|:---------------------------------:|-----------------------------------------------------------------------------------------------|
+| BFT Time  | Guaranteed        | Tolerates `< 1/3` Byzantine nodes | Best effort and **not guaranteed**.                                                           |
+| PBTS      | Guaranteed        | Tolerates `< 2/3` Byzantine nodes | Guaranteed with `ACCURACY` determined by the consensus parameters `PRECISION` and `MSGDELAY`. |
```
Comment (adizere) [inline spec/consensus/time.md:53]: Really good overview!

File: spec/consensus/time.md [hunk 5]
```diff
@@ -33,20 +37,53 @@ CometBFT implements two algorithms for computing block times:
...
+## Adopting PBTS
+
+The Proposer-Based Timestamp (PBTS) algorithm is the recommended algorithm for
+producing block times.
+
+As of CometBFT `v1.x`, however, PBTS is not enabled by default, either for new
+chains using default values for genesis parameters, nor for chains upgrading to
+newer CometBFT versions, for backwards compatibility reasons.
+
+Enabling PBTS requires configuring some consensus parameters:
```
Comment (adizere) [inline spec/consensus/time.md:70]: Please consider adding a link here to a reference for _consensus params_ (either docs.cometbft.com or spec/core/params.md).

File: spec/core/params.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+The feature enable height parameters cannot be set to heights lower or equal to
+the current blockchain height.
+
+Once a feature enable height is set and reached by the blockchain, the feature
+enable height cannot be updated.
+
+## Initial Parameters
+
+TODO: genesis
```
Comment (adizere) [inline spec/core/params.md:193]: It's fine to just put a link to docs.cometbft.com genesis reference here potentially

File: spec/core/params.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+no consensus parameter is updated.
+
+Otherwise, each field that is set (not empty) is applied in full, namely, every
+parameter defined in that category of consensus parameters is updated.
+For example, if the `BlockParams` field is set because the application wants to
+update the configured `MaxBytes` consensus parameter, the application MUST also
+set every other parameter of `BlockParams` (e.g., `MaxGas`), even if the
+application does not want to update the value of those parameters.
+
+If the application does not set values for every parameter of a category of
```
Comment (adizere) [inline spec/core/params.md:212]:
```suggestion
**Important:** If the application does not set values for every parameter of a category of
```

Comment (adizere) [reply spec/core/params.md:212]: Seems like a very important remark.

File: docs/explanation/core/proposer-based-timestamps.md [hunk 8]
```diff
@@ -55,44 +33,176 @@ The following protocols and application features require a reliable source of ti
...
+Enabling PBTS requires configuring the [consensus parameters](#consensus-parameters)
+that govern the operation of the algorithm.
+
+### Selecting a Timestamp
 
-When the proposer node creates a new block proposal, the node reads the time
-from its local clock and uses this reading as the timestamp for the proposed
-block.
+When a validator creates a new block, it reads the time from its local clock
+and uses this reading as the timestamp for the block.
```
Comment (adizere) [inline docs/explanation/core/proposer-based-timestamps.md:51]:
```suggestion
When a validator creates a new block to propose it, it reads the time from its local clock
and uses this reading as the timestamp for the proposed block.
```

nit

File: docs/explanation/core/proposer-based-timestamps.md [hunk 9]
```diff
@@ -55,44 +33,176 @@ The following protocols and application features require a reliable source of ti
...
+should periodically synchronize their clocks to a reliable NTP server.
+Validators whose clocks drift too far away from the rest of the network will no
+longer propose blocks with valid timestamps. Additionally, they will not consider
+the timestamps of blocks proposed by their peers to be valid either.
+
+
+## Consensus Parameters
+
+The functionality of the PBTS algorithm is governed by two consensus
+parameters: the synchronous parameters `Precision` and `MessageDelay`.
```
Comment (adizere) [inline docs/explanation/core/proposer-based-timestamps.md:92]:
```suggestion
parameters: the synchrony parameters `Precision` and `MessageDelay`.
```

nit

File: docs/explanation/core/proposer-based-timestamps.md [hunk 10]
```diff
@@ -55,44 +33,176 @@ The following protocols and application features require a reliable source of ti
...
+An additional consensus parameter `PbtsEnableHeight` is used to enable PBTS
+when instantiating a new network or when upgrading an existing a network that
+uses BFT Time.
+
```
Comment (adizere) [inline docs/explanation/core/proposer-based-timestamps.md:96]:
```suggestion
when instantiating a new network or when upgrading an existing network that
uses currently BFT Time.

```

File: docs/explanation/core/proposer-based-timestamps.md [hunk 11]
```diff
@@ -55,44 +33,176 @@ The following protocols and application features require a reliable source of ti
...
+In heights up to `PbtsEnableHeight - 1`, the generation and validation of block
+timestamps in the network will be carried out using the legacy BFT Time algorithm.
+
+While `PbtsEnableHeight` is set to `0`, the network will adopt the legacy BFT
+Time algorithm.
+
+When `PbtsEnableHeight` is set to a height `H > 0`, the network will switch to
+the PBTS algorithm from height `H`.
+The network will still adopt the legacy BFT Time algorithm to produce and
+validate block timestamps until height `H - 1`.
```
Comment (adizere) [inline docs/explanation/core/proposer-based-timestamps.md:141]:
```suggestion
validate block timestamps until height `H - 1` (including that height).
```

**Review by adizere — APPROVED** [review]
Awesome work Daniel! Left some suggestions, let's fix the open TODOs and remarks and GTM.

File: spec/consensus/time.md (hunk 5 above)
Comment (cason) [reply spec/consensus/time.md:70]: This file is outdated, but commenting here to not forget to consider.

File: proto/cometbft/types/v1/params.proto (hunk 2 above)
Comment (cason) [reply proto/cometbft/types/v1/params.proto:95]: Outdated too, double checking after merge.

File: proto/cometbft/types/v1/params.proto (hunk 1 above)
Comment (cason) [reply proto/cometbft/types/v1/params.proto:54]: Probably yes, I'll look into.

**Comment by andynog** [comment]
@cason is this ready to be reviewed or still wip ?

**Comment by cason** [comment]
WIP, it is incomplete.

**Pull request closed without merging**